ocrd/validator/page_validator.py

```diff
@@ -0,0 +1,174 @@
+import re
+
+from ocrd.utils import getLogger
+from ocrd.model.ocrd_page import from_file, parse
+
+from ocrd.model.ocrd_page_generateds import (
+    PcGtsType,
+    PageType,
+    TextEquivType,
+    TextRegionType,
+    TextLineType,
+    WordType,
+    GlyphType,
+)
+from .report import ValidationReport
+
+log = getLogger('ocrd.page_validator')
+
+_HIERARCHY = [
+    (PageType,       'Page',       'get_TextRegion', None), # pylint: disable=bad-whitespace
+    (TextRegionType, 'TextRegion', 'get_TextLine',   '\n'), # pylint: disable=bad-whitespace
+    (TextLineType,   'TextLine',   'get_Word',       ' '),  # pylint: disable=bad-whitespace
+    (WordType,       'Word',       'get_Glyph',      ''),   # pylint: disable=bad-whitespace
+    (GlyphType,      'Glyph',      None,             None), # pylint: disable=bad-whitespace
+]
+
+class ConsistencyError(Exception):
+
+    def __init__(self, tag, ID, actual, expected):
+        self.tag = tag
+        self.ID = ID
+        self.actual = actual
+        self.expected = expected
+        super(ConsistencyError, self).__init__("INCONSISTENCY in %s ID '%s': text results '%s' != concatenated '%s'" % (tag, ID, actual, expected))
+
+class PageValidator(object):
+
+    @staticmethod
+    def validate_filename(filename, **kwargs):
```

Comment:
I do not like the naming scheme here: `validate_filename` is misleading especially given its counterpart `validate_ocrd_page`.

Reply:
The naming scheme is `validate_<source-of-the-page-xml>`. `validate_ocrd_page` means that the source is an `OcrdPage` object, `validate_filename` means PAGE XML comes from a file on disk.

What would you prefer? `validate_by_filename` or similar?

Reply:
Difficult. Is it technically possible to have a `validate` method which calls the appropriate method depending on the parameter type?